Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -50,127 +50,27 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
-        size_t bindingOffsetForStage[] = { 0, 0, 0 };
-        for (uint32_t i = 0; i < entryCount; ++i) {
-            const WGPUBindGroupEntry& entry = descriptor.entries[i];
-
-            if (entry.nextInChain)
-                return BindGroup::createInvalid(*this);
-
-            bool bufferIsPresent = WebGPU::bufferIsPresent(entry);
-            bool samplerIsPresent = WebGPU::samplerIsPresent(entry);
-            bool textureViewIsPresent = WebGPU::textureViewIsPresent(entry);
-            if (bufferIsPresent + samplerIsPresent + textureViewIsPresent != 1)
-                return BindGroup::createInvalid(*this);
-
-            auto stages = bindGroupLayout ? bindGroupLayout->stagesForBinding(entry.binding) : (WGPUShaderStage_Vertex | WGPUShaderStage_Fragment | WGPUShaderStage_Compute);
-            for (size_t currentStage = 0; currentStage < shaderStageLength; ++currentStage) {
-                WGPUShaderStage renderStage = shaderStage[currentStage];
-                if (!(stages & renderStage))
-                    continue;
-
-                auto& offset = bindingOffsetForStage[currentStage];
-                ASSERT(argumentBufferContents[currentStage]);
-
-                if (bufferIsPresent) {
-                    id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
-                    if (entry.offset > buffer.length)
-                        return BindGroup::createInvalid(*this);
-
-                    ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
-                    *(float**)(argumentBufferContents[currentStage] + offset) = (float*)((char*)buffer.gpuAddress + entry.offset);
-                    offset += sizeof(MTLResourceID);
-                } else if (samplerIsPresent) {
-                    id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
-                    ASSERT(sizeof(float*) == sizeof(sampler.gpuResourceID));
-                    *(MTLResourceID*)(argumentBufferContents[currentStage] + offset) = sampler.gpuResourceID;
-                    offset += sizeof(MTLResourceID);
-                } else if (textureViewIsPresent) {
-                    id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
-                    ASSERT(sizeof(float*) == sizeof(texture.gpuResourceID));
-                    *(MTLResourceID*)(argumentBufferContents[currentStage] + offset) = texture.gpuResourceID;
-                    resources.append({ texture, MTLResourceUsageRead, renderStage });
-                    offset += sizeof(MTLResourceID);
-                }
-            }
-        }
 
-        return BindGroup::create(argumentBuffer[0], argumentBuffer[1], argumentBuffer[2], WTFMove(resources), *this);
+    id<MTLArgumentEncoder> argumentEncoder[] = { bindGroupLayout.vertexArgumentEncoder(), bindGroupLayout.fragmentArgumentEncoder(), bindGroupLayout.computeArgumentEncoder() };
+    for (size_t i = 0; i < shaderStageLength; ++i) {
```

Comment:
How do you feel about naming these values of `i` in an enum, and doing this in a lambda instead of a loop? `EnumeratedArray` is a pretty elegant thing to use for situations like these.

```
enum class ShaderType {
    Vertex,
    Fragment,
    Compute
};
EnumeratedArray<ShaderType, id<MTLArgumentEncoder>, ShaderType::Compute> argumentEncoders;
auto foo = [](ShaderType type) {
    ... argumentEncoders[type] ...
}
foo(ShaderType::Vertex);
foo(ShaderType::Fragment);
foo(ShaderType::Compute);
```

Reply:
This is a good suggestion and would be easier if we didn't have the nested loop over entries where we check `bufferIsPresent`, `samplerIsPresent`, etc. Unless I misunderstand the suggestion, applying it (naively at least) would result in `3 * entryCount` of these checks whereas currently we have only `entryCount` checks, not to mention potentially worse data access patterns assuming entryCount > 3.

Reply:
I think using EnumeratedArray makes sense, I'm attempting to apply the suggestion of using EnumeratedArray  without using the lambdas since it doubles the code lines of the first for loop which I think impacts readability.